File: tests/runtime/resource-import-and-export/intermediate.cs
using IntermediateWorld.wit.Imports.test.resourceImportAndExport;
using Import = IntermediateWorld.wit.Imports.test.resourceImportAndExport.ITestImports;

namespace IntermediateWorld.wit.Exports.test.resourceImportAndExport
{
    public class TestExportsImpl : ITestExports {
	public class Thing : ITestExports.Thing, ITestExports.IThing {
	    public Import.Thing val;

	    public Thing(uint v) {
		this.val = new Import.Thing(v + 1);
	    }

	    public uint Foo() {
		return this.val.Foo() + 2;
	    }

	    public void Bar(uint v) {
		this.val.Bar(v + 3);
	    }

	    public static ITestExports.Thing Baz(ITestExports.Thing a, ITestExports.Thing b) {
		return new Thing(Import.Thing.Baz(((Thing) a).val, ((Thing) b).val).Foo() + 4);
	    }
	}
    }
}

namespace IntermediateWorld {
    public class IntermediateWorldExportsImpl : ITestImports
    {
	public static Import.Thing ToplevelExport(Import.Thing things) {
	    return IntermediateWorld.IIntermediateWorldImports.ToplevelImport(things);
	}
    }
}

File: tests/runtime/resource-import-and-export/intermediate.rs
include!(env!("BINDINGS"));

use exports::test::resource_import_and_export::test::{GuestThing, Thing as ExportThing};
use std::cell::RefCell;

pub struct Test {}

export!(Test);

pub struct MyThing {
    thing: RefCell<Option<Thing>>,
}

impl Guest for Test {
    fn toplevel_export(input: Thing) -> Thing {
        toplevel_import(input)
    }
}

impl exports::test::resource_import_and_export::test::Guest for Test {
    type Thing = MyThing;
}

impl GuestThing for MyThing {
    fn new(v: u32) -> Self {
        Self {
            thing: RefCell::new(Some(Thing::new(v + 1))),
        }
    }

    fn foo(&self) -> u32 {
        let thing = self.thing.borrow();
        let thing = thing.as_ref().unwrap();
        thing.foo() + 2
    }

    fn bar(&self, v: u32) {
        let mut thing = self.thing.borrow_mut();
        let thing = thing.as_mut().unwrap();
        thing.bar(v + 3);
    }

    fn baz(a: ExportThing, b: ExportThing) -> ExportThing {
        let mut a = a.get::<MyThing>().thing.borrow_mut();
        let mut b = b.get::<MyThing>().thing.borrow_mut();
        let result =
            Thing::baz(Option::take(&mut a).unwrap(), Option::take(&mut b).unwrap()).foo() + 4;
        ExportThing::new(MyThing::new(result))
    }
}
